### PR TITLE
[FLINK-5874] Restrict key types in the DataStream API.

### PR DESCRIPTION
**Pull request by kl0u**
Rejects a type from being a key in `DataStream.keyBy()` if it is:
1. it is a POJO type but does not override the `hashCode()` and
   relies on the `Object.hashCode()` implementation.
2. it is an array of any type.

This was also discussed with @fhueske 

R @aljoscha

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 1]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
...
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+		expectedException.expectMessage("This type (" + expectedTypeInfo + ") cannot be used as key.");
+
+		input.keyBy("id");
+	}
+
+	@Test
+	public void testNestedArrayWorkArround() {
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: type: Arround -> Around

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 2]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
...
+		input.keyBy(new KeySelector<Tuple2<Integer[],String>, Tuple2<Integer[],String>>() {
+			@Override
+			public Tuple2<Integer[], String> getKey(Tuple2<Integer[], String> value) throws Exception {
+				return value;
+			}
+		});
+	}
+
+	@Test
+	public void testPrimitiveKeyRejection() throws Exception {
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: what do we actually verify here?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 3]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
 	}
 
 	/////////////////////////////////////////////////////////////
+	// KeyBy testing
+	/////////////////////////////////////////////////////////////
+
+	@Rule
+	public ExpectedException expectedException = ExpectedException.none();
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: First time I'm seeing this pattern and i can't help but feel that a simple try-catch block would be more readable.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java [hunk 4]
```diff
@@ -736,7 +736,7 @@ public void restoreState(List<Integer> state) throws Exception {
 
 		static final ValueStateDescriptor<Boolean> descriptor = new ValueStateDescriptor<>("seen", Boolean.class, false);
 		private static final long serialVersionUID = 1L;
-		private ValueState<Boolean> operatorState;
+		private transient ValueState<Boolean> operatorState;
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java]: unrelated change.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 5]
```diff
@@ -114,9 +121,53 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
+
+	private TypeInformation<KEY> validateKeyType(TypeInformation<KEY> keyType) {
+		Stack<TypeInformation<?>> stack = new Stack<>();
+		stack.push(keyType);
+
+		while (!stack.isEmpty()) {
+			TypeInformation<?> typeInfo = stack.pop();
+
+			if (!validateKeyTypeIsHashable(typeInfo)) {
+				throw new InvalidProgramException("This type (" + keyType + ") cannot be used as key.");
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: Let's include ```typeInfo``` in this exception as well to narrow it down for the user. We may even want to delay the exception until we scanned the entire type and report all invalid keys at once.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 6]
```diff
@@ -114,9 +121,53 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
+	 *     <li>it is a POJO type but does not override the {@link #hashCode()} method and relies on
+	 *     the {@link Object#hashCode()} implementation.</li>
+	 *     <li>it is an array of any type (see {@link PrimitiveArrayTypeInfo}, {@link BasicArrayTypeInfo},
+	 *     {@link ObjectArrayTypeInfo}).</li>
+	 * </ol>,
+	 * {@code true} otherwise.
+	 */
+	private boolean validateKeyTypeIsHashable(TypeInformation<?> type) {
+		try {
+			return (type instanceof PojoTypeInfo) ?
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: ? and : are typically at the start of a new line.

Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: we should also simply break this if statement down into multiple blocks.

i.e.

```
if (type isntanceof PojoTypeInfo) {
return //find hashCode
}

if (tpye instance off XArrayTypeInfo ... ) {
 return false;
}

return true;
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 7]
```diff
@@ -114,9 +121,53 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
+					!type.getTypeClass().getMethod("hashCode").getDeclaringClass().equals(Object.class) :
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: this method would be more readable by not inverting here, and returning true as the default.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java [hunk 8]
```diff
@@ -114,9 +121,53 @@ public KeyedStream(DataStream<T> dataStream, KeySelector<T, KEY> keySelector, Ty
...
+			}
+		}
+		return keyType;
+	}
+
+	/**
+	 * Validates that a given type of element (as encoded by the provided {@link TypeInformation}) can be
+	 * used as a key in the {@code DataStream.keyBy()} operation.
+	 *
+	 * @return {@code false} if:
```
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: I would shorten this to read ```returns true if the type overrides the hashcode implementation```. The details can be container in the general javadoc of the method.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 9]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
...
+	private static class TestClass {
+
+		private final int id;
+
+		TestClass(int id) {
+			this.id = id;
+		}
+	}
+
+	public static class POJOnoHashCode {
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: no being lower case appears to violate the naming convention.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 10]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
...
+			}
+		};
+
+		testKeyRejection(keySelector, BasicArrayTypeInfo.INT_ARRAY_TYPE_INFO);
+	}
+
+	@Test
+	public void testObjectArrayKeyRejection() {
+
+		KeySelector<Tuple2<Integer[], String>, TestClass[]> keySelector =
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: Do we need a separate class here or could we just throw in an Object[]?

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java (hunk 2 above)
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: The name should be changed; as it stands i expected that this test verifies that a primitive key is rejected.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 11]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
...
+		public int[] getId() {
+			return id;
+		}
+
+		public void setId(int[] id) {
+			this.id = id;
+		}
+	}
+
+	public static class POJOwithHashCode {
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: You could extend PojoNoHashCode and only implement hashCode.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java (hunk 9 above)
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: the naming is also inconsistent with the other class; this class name doesn't contain "with".

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 12]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
...
+		DataStream<POJOnoHashCode> input = env.fromElements(
+				new POJOnoHashCode(new int[] {1, 2}));
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+
+		input.keyBy(keySelector);
+	}
+
+	// composite key tests : Tuples
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: this commit will not be seen as it is.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 13]
```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
...
+		expectedException.expectMessage("This type (" + expectedTypeInfo + ") cannot be used as key.");
+
+		input.keyBy("id");
+	}
+
+	@Test
+	public void testNestedArrayWorkArround() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<POJOwithHashCode> input = env.fromElements(
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: The focus of this test appears to be to verify that you can use an array as a key by wrapping it in a Pojo that implements HashCode. (based on the naming).

We should probably focus more on the Pojo-with-hashCode implementation part; doesn't really matter what is contained. We are never checking that anyway.

**Comment by zentol** [comment]
This should work and is well tested, good job. Had a bunch of minor comments, but nothing critical.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java (hunk 4 above)
Comment (kl0u) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java]: true but it is wrong and someone has to fix it at some point.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java (hunk 8 above)
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: I think it is worth having it also here for users of the method. The more the places that the user can find the required information, the better.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java (hunk 6 above)
Comment (kl0u) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: Really? I am not sure about it.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java (hunk 3 above)
Comment (kl0u) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: This is the proper way of doing it. A `try catch` is of course an option, but this is more concise, and built in JUnit.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java (hunk 10 above)
Comment (kl0u) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: We can go for `Object[]`.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java (hunk 2 above)
Comment (kl0u) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: The name is totally misleading, you are right. This test only verifies that we did not break anything and that even if you leave the default signature for the `keySelector` (returning `Object` although your key is of a different type) it still works.

**Comment by kl0u** [comment]
Thanks for the thorough review @zentol . I integrated the comments that I agree with and commented on the rest. If you agree with the changes, I can merge it.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java (hunk 4 above)
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java]: well feel free to open a new PR or create a hotfix.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java (hunk 6 above)
Comment (zentol) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/KeyedStream.java]: So i looked through the code base, and found 5 different patterns, so you're right, "typical" doesn't really fit:
1:
```
<condition> ? <then> : <else>
```
2:
```
<condition> ? 
 <then> : <else>
```
3:
```
<condition> ?
 <then>
 : <else>
```
4:
```
<condition>
 ? <then>
 : <else>
```
5:
```
if <condition> ?
 <then> :
 <else>
```

In my opinion, only 1 and 4 should be used; and 1 only for short statements. Both of these options are symmetrical; the if and else statement are formatted the same way, and place important information (? and \:) at the start of the line.

The point here is pattern recognition.
When seeing this
```
X
 ? Y
 : Z
```
you can immediately see what is going on without looking at the details of the expression, nor scanning any line; the very first character gives it away. The position of ? and \: is also static, it's on the next line with an additional indent, compared to other options which may hide ? at the end of a 100 character expression (which gets especially funky with option 2 which also hides \: somewhere, so at a glace it could also just be a method call with many arguments).

**Comment by kl0u** [comment]
Done! Let me know if you have any additional comments @zentol

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java (hunk 3 above)
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: Well it may be more concise, but a try catch block makes it way more obvious that this part of the code throws an exception we care about.

try-catch blocks are also more commonin the codebase, and I'm wondering whether we want to introduce an inconsistency there.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 14]
```diff
@@ -906,6 +919,243 @@ public void testChannelSelectors() {
...
+
+	@Test
+	public void testObjectArrayKeyRejection() {
+
+		KeySelector<Tuple2<Integer[], String>, Object[]> keySelector =
+				new KeySelector<Tuple2<Integer[], String>, Object[]>() {
+
+					@Override
+					public Object[] getKey(Tuple2<Integer[], String> value) throws Exception {
+						Object[] ks = new TestClass[value.f0.length];
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: Can we use plain objects here? Then we wouldn't need the TestClass at all.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java [hunk 15]
```diff
@@ -906,6 +919,243 @@ public void testChannelSelectors() {
...
+		});
+
+		keyedStream.addSink(new SinkFunction<Integer>() {
+			@Override
+			public void invoke(Integer value) throws Exception {
+				Assert.assertEquals(10000L, (long) value);
+			}
+		});
+
+		env.execute();
```
Comment (zentol) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: We don't need to execute a job, correct? just the keyBy passing should be enough.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java (hunk 3 above)
Comment (kl0u) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java]: The thing is that if I remember correctly we discussed about how to check for exceptions earlier this year. And if I remember correctly the outcome was sth along the lines of using JUnit facilities. Feel free to correct me on this one.

**Comment by zentol** [comment]
+1 to merge.
